service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/gcp/dataproccluster/CreateDataprocClusterStep.java

```diff
@@ -0,0 +1,354 @@
+    Cluster cluster =
+        createClusterModel(
+            flightContext,
+            projectId,
+            resource.getClusterId(),
+            petEmail,
+            workspaceUserFacingId,
+            creationParameters,
+            stagingBucket.getBucketName(),
+            tempBucket.getBucketName(),
+            cliConfiguration.getServerName());
+
+    DataprocCow dataprocCow = crlService.getDataprocCow();
+    try {
+      OperationCow<Operation> creationOperation;
+      try {
+        creationOperation =
+            dataprocCow
+                .regionOperations()
+                .operationCow(dataprocCow.clusters().create(clusterName, cluster).execute());
+      } catch (GoogleJsonResponseException e) {
+        // If the cluster already exists, this step must have already run successfully. Otherwise
+        // retry.
+        if (HttpStatus.CONFLICT.value() == e.getStatusCode()) {
+          logger.debug("Dataproc cluster {} already created.", clusterName.formatName());
+          return StepResult.getStepResultSuccess();
+        } else if (HttpStatus.BAD_REQUEST.value() == e.getStatusCode()) {
+          // Don't retry bad requests, which won't change. Instead, fail faster.
+          return new StepResult(StepStatus.STEP_RESULT_FAILURE_FATAL, e);
+        }
+        return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+      }
+
+      GcpUtils.pollAndRetry(creationOperation, Duration.ofSeconds(20), Duration.ofMinutes(15));
+    } catch (IOException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+    return StepResult.getStepResultSuccess();
+  }
+
+  private static Cluster createClusterModel(
+      FlightContext flightContext,
+      String projectId,
+      String clusterId,
+      String serviceAccountEmail,
+      String workspaceUserFacingId,
+      ApiGcpDataprocClusterCreationParameters creationParameters,
+      String stagingBucketName,
+      String tempBucketName,
+      String cliServer) {
+    Cluster cluster = new Cluster();
+    setFields(
+        clusterId,
+        creationParameters,
+        stagingBucketName,
+        tempBucketName,
+        serviceAccountEmail,
+        workspaceUserFacingId,
+        cliServer,
+        cluster);
+    setNetworks(cluster, projectId, flightContext.getWorkingMap());
+    return cluster;
+  }
+
+  @VisibleForTesting
+  static Cluster setFields(
+      String clusterId,
+      ApiGcpDataprocClusterCreationParameters creationParameters,
+      String stagingBucketName,
+      String tempBucketName,
+      String serviceAccountEmail,
+      String workspaceUserFacingId,
+      String cliServer,
+      Cluster cluster) {
+
+    cluster
+        .setClusterName(clusterId)
+        .setConfig(
+            new ClusterConfig()
+                .setConfigBucket(stagingBucketName)
+                .setTempBucket(tempBucketName)
+                // TODO PF-2828: Add post-startup script
```

Comment:
Can this read from the value passed in `creationParameters`, even if we don't have a default script yet?
Also given how important this is to the Notebook case, is it worth calling this out as a required followup in the PR description?

Reply:
Yes and yes.